### ISSUE TITLE
In (D)TLS engine ERR_clear_error() calls should be avoided

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#10021]
As (partly) discussed for PR #8716, the code in `ssl/` contains many needless occurrences of `ERR_clear_error()`.  This is pretty unfortunate in case higher-level (library or application) code has produced beforehand some error queue entries that should possibly be printed later, after calls to SSL/(D)TLS code, as helpful diagnostics.

In many of these occurrences the intention apparently is just to remove any (D)TLS-related additions to the error queue in case intermediate lower-level errors need to be ignored, yet they do too much: they kill everything in the error queue. In such cases, a combination of `ERR_set_mark()` and `ERR_pop_to_mark()` should be used instead.

For other such occurrences the intention supposedly is to flag internally that no (D)TLS error has occurred. Instead of checking the error queue, this should be detected in other ways, such as checking `s->statem.state` for (non-)error states.

**Comment by mattcaswell** [comment]
It is an unfortunate aspect of the design of libssl that SSL_get_error() checks the error queue as one of the first things it does to figure out what kind of error happens:

https://github.com/openssl/openssl/blob/a6dd3a3aa2577fa6c3586b96fe0b46be336cff54/ssl/ssl_lib.c#L3623-L3641

If we were to replace all the `ERR_clear_error()` calls with `ERR_set_mark()`/`ERR_pop_to_mark()` then SSL_get_error() will return the wrong result if SSL I/O functions are called with stale errors on the queue.

So, in order to do as you suggest, we need to refactor `SSL_get_error` to be more intelligent. As I mentioned in #8716 it *might* be sufficient to check if `s->statem.state == MSG_FLOW_ERROR`

**Comment by levitte** [comment]
I'm looking at this now, and I can't see a reason why we should keep errors that are, for all intents and purposes, *stale* at this point in the code path.  @DDvO, is this possibly related to the trick we've talked about recently, where you pre-insert an error in the error queue just to decide later if you want it to stick around or not?  Consider what we concluded in that discussion, this issue might not be current any more...

**Comment by DDvO** [comment]
Good point to check whether the motivation for this issue is gone after the outcome of #10537.
I found the following.

* The change to stop using tentative error queue entries for building up diagnostic info on the search for a suitable sender cert of a CMP message only partially  solves the issue I brought up here because - though part of the information is now sent to the log rather than to the error queue - some final higher-level error queue entry is produced now in the usual manner and this entry is still threatened/killed by subsequent calls to the SSL engine.

* There was a second case where the CMP client tentatively produced an error queue entry, namely before establishing a TLS connection, in order to be able to add more specific error information if/when it becomes available in the course of connection setup. Yet this can be solved in the same way as discussed in #10537, and it turns out that (in the course of carving out the HTTP(S) client code as discussed in #10271) I've already eliminated this.

* In case the CMP client has received a response that it rejects, producing some specific error queue entries with more detail on the rejection (which can be independent of the types of errors I referred to above), the client typically sends negative feedback also to the server. So it uses the SSL engine again to send a CMP error message. During this operation the original error queue entries are typically discarded, and so the higher-level cert management application cannot learn them by consulting the error queue after the CMP client returns.

The latter unfortunate situation was the main reason for me to file this issue. 
My current workaround is that the CMP client lib code dumps the error queue entries (using, e.g., `ERR_print_errors_fp(stderr)` or the CMP logging facility) _before_ using the SSL engine to send out the CMP error message. It would be better though if the application above the library obtained those error queue entries and could decide itself when and how to inspect/print/handle them.

I'm pretty sure that there are similar situations also for other higher-level protocols and applications.

**Comment by DDvO** [comment]
Two points independent of concrete scenarios where the loss of error queue entries hurts.
* As pointed out above by @mattcaswell:
it is unfortunate that part of the SSL engine relies on the error queue being empty.
* I add here a more general point: 
a well-designed lib/component does not discard error queue entries it is not responsible for.

**Comment by levitte** [comment]
>  a well-behaved lib/component does not discard error queue entries it is not responsible for.

Depends on the basic assumptions made.  One I'm making is that a well behaved application has checked for errors so far, or chosen to ignore them so far.  So the basic assumption is that anything hanging around in the error queue is unrelated and therefore uninteresting, and actually leaving entries in the error queues gives false reports on what went wrong.

**Comment by levitte** [comment]
It may be that we have a confusion on what the error queue represents.  Is it a record of the *current* error, or is it a log?  I've always thought of it as the former, so...

**Comment by mattcaswell** [comment]
> It may be that we have a confusion on what the error queue represents. Is it a record of the current error, or is it a log? I've always thought of it as the former, so...

Well...define "current". In @DDvO's third scenario above the "current" error is in relation to some broader transaction that is occurring that may result in multiple interactions with libssl. So the errors needs to stick around until the upper layers of the code are ready to deal with them. This doesn't seem unreasonable.

**Comment by DDvO** [comment]
My assumption is that the OpenSSL error queue should convey to the calling application details on all (relevant) errors that occurred during the call.
The SSL portion of OpenSSL cannot tell and should not decide what is relevant to the application and throw away anything else. 

The problem I mention here only surfaces in more complex/high-level scenarios where (part of) OpenSSL, in this case the crypto lib, makes (indirect) use of the SSL lib and there are intermediate errors that should be preserved beyond calls to the SSL engine before they can be returned to the application.

**Comment by nhorman** [comment]
@DDvO so is the ask here to alter calls, like in SSL_CTX_new_ex

```
{
        ENGINE *eng;
        eng = ENGINE_by_id(eng_str(OPENSSL_SSL_CLIENT_ENGINE_AUTO));
        if (!eng) {
            ERR_clear_error();
            ENGINE_load_builtin_engines();
            eng = ENGINE_by_id(eng_str(OPENSSL_SSL_CLIENT_ENGINE_AUTO));
        }
        if (!eng || !SSL_CTX_set_client_cert_engine(ret, eng))
            ERR_clear_error();
    }
```

To remove the call to ERR_clear_error, and let the application interrogate them as they see fit?  It makes sense on paper, but I could see that causing significant problems for applications in the field

**Comment by DDvO** [comment]
As I wrote in the OP, those 'brute-force' `ERR_clear_error()` calls should be replaced by suitable combinations of `ERR_set_mark()` and `ERR_pop_to_mark()`, such that the SSL lib only removes those error queue entries it added itself.
Already pre-existing error queue entries from other parties should be retained, and it is the responsibility of those parties what they add to the queue and potentially remove.

Why do you see potentially significant problems with this with existing apps?
Every caller of an OpenSSL function should be prepared to cope with any number and sort of error queue entries on failure of the given function call. Usually, they anyway give up on error, and some outer/cleanup layer then has the chance to provide diagnostic data from the error queue entries that so far got killed by libssl. 
When the apps call other parts of OpenSSL, which typically do not wipe the error queue, they already can handle this, and so I see not much risk that they cannot cope with pre-existing queue elements not being wiped in the future.

**Comment by nhorman** [comment]
If you're talking about using pop_to_mark, then this seems reasonable to me
